components/script/dom/bindings/structuredclone.rs

```diff
@@ -46,15 +47,18 @@ impl StructuredCloneData {
         })
     }
 
-    /// Converts a StructuredCloneData to Vec<u64>
-    pub fn move_to_arraybuffer(self) -> Vec<u64> {
+    /// Converts a StructuredCloneData to Vec<u8> for inter-thread sharing
+    pub fn move_to_arraybuffer(self) -> DOMMessage {
         unsafe {
-            slice::from_raw_parts(self.data, self.nbytes).to_vec()
+            DOMMessage(slice::from_raw_parts(self.data as *mut u8, self.nbytes).to_vec())
         }
     }
 
-    /// Converts back to StructuredCloneData using a pointer and no of bytes
-    pub fn make_structured_clone(data: *mut u64, nbytes: size_t) -> StructuredCloneData {
+    /// Converts back to StructuredCloneData
+    pub fn make_structured_clone(data: DOMMessage) -> StructuredCloneData {
+        let DOMMessage(mut data) = data;
+        let nbytes = data.len();
+        let data = data.as_mut_ptr() as *mut u64;
         StructuredCloneData {
             data: data,
```

Comment:
This is unsafe - the actual Vec value goes out of scope when we return, so the `StructuredCloneData` contains a pointer to freed memory. We should make `StructuredCloneData` an enum that can store a backing `Vec` if necessary.